### PR TITLE
SAST controls autoupdate for SDARQ - DSEC-585

### PR DESCRIPTION
**Pull request by TomConner**
Update sast-details collection from Codacy, SonarCloud, and GitHub. Then update security-controls collection according to the sast-details data.

Adds secret tokens to access

### CODE REVIEW

File: security-controls/sast.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+import logging
+import os
+import re
+
+import requests
+from google.cloud import firestore
+from requests.auth import HTTPBasicAuth
+
+SECURITY_CONTROLS = os.environ['SC_FIRESTORE_COLLECTION']
+SAST_DETAILS = os.environ['SAST_FIRESTORE_COLLECTION']
```
Comment (zbedo) [inline security-controls/sast.py:12]: If the information stored in that collection does not help the cronjob to run correctly, I don't think we should store more information. In case we need that information in the future for another task, then we can add it again.

File: docs/src/docs/secret_deplyment.md [hunk 2]
```diff
@@ -44,6 +44,36 @@ metadata:
...
+```
+Name:         codacy
+Namespace:    security-controls
+Type:  Opaque
+
+Data
+====
+CODACY_API_KEY:  Codacy API key
+```
+
```
Comment (zbedo) [inline docs/src/docs/secret_deplyment.md:76]: Is there any reason why SONARCLOUD_API_KEY, GITHUB_TOKEN, and CODACY_API_KEY are not in the same secret file `security-controls`?

Comment (zbedo) [reply docs/src/docs/secret_deplyment.md:76]: All `security-controls` namespace secrets are in the same file. Looks like you put these 3 in different files.

File: sdarq/deploy.sh [hunk 3]
```diff
@@ -24,6 +24,7 @@ export SDARQ_SECRET="sdarq"
 export CIS_JOB_TOPIC="cis-scans"
 export ZAP_JOB_TOPIC="zap-scans"
 export SC_FIRESTORE_COLLECTION="security-controls"
+export SAST_FIRESTORE_COLLECTION="sast-details"
```
Comment (zbedo) [inline sdarq/deploy.sh:27]: Why `SAST_FIRESTORE_COLLECTION` is added to SDARQ? SDARQ's `frontend` or `backend` is not using that Firestore collection.

File: security-controls/sast.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    list_github(repos)
+    return repos
+
+def update_sast_values():
+    '''Update security-controls and sast-details in Firestore.'''
+
+    try:
+        logging.info("update_sast_values")
+
+        logging.debug("lengths %s %s %s",
```
Comment (zbedo) [inline security-controls/sast.py:278]: Why are we debugging the length of Codacy, Sonarcloud and Github token?

File: security-controls/sast.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+            org = 'DataBiosphere'
+        if org.lower() == 'broadinstitute':
+            org = 'broadinstitute'
+        repo = get_repo(repos, (org, repo_name))
+        sc_record[ID] = doc.id
+        repo[SC_PREVIOUS] = sc_record
+        logging.debug('SAST controls to be updated for %s github="%s"', doc.id, repo_url)
+
+
+def list_github(repos: Repos):
```
Comment (zbedo) [inline security-controls/sast.py:96]: Why do we need Github metadata for all repos?

**Comment by zbedo** [comment]
For the `sast-details` collection that is added, as long as we don't need that information in this SAST update in security controls, I don't think we should include it in this PR. In the future, if we will need all that information we can add it again.

File: security-controls/deployment.yaml [hunk 6]
```diff
@@ -36,11 +37,19 @@ spec:
             envFrom:
             - secretRef:
                 name: ${SERVICE_SECRET}
+            - secretRef:
```
Comment (zbedo) [inline security-controls/deployment.yaml:40]: Can we add these secrets in the same file with other `security-controls` namespace secrets?

File: docs/src/docs/secret_deplyment.md (hunk 2 above)
Comment (zbedo) [reply docs/src/docs/secret_deplyment.md:76]: Can we please add these secrets following the documentation format?

File: security-controls/sast.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+RE_GROUP_ORG = 2
+RE_GROUP_REPO = 3
+
+SONAR_ORGS = {
+    "broadinstitute": "dsp-appsec",
+    "DataBiosphere": "broad-databiosphere",
+    "CancerDataAggregator": "cancerdataaggregator"
+}
+
+CODACY_ORGS = [
```
Comment (zbedo) [inline security-controls/sast.py:52]: Can we set this in cloud build variables, so we don't have hardcoded values in the code.

File: security-controls/sast.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+def codacy_org(organization: str) -> str:
+    '''path for codacy org'''
+    return f'{CODACY_BASE}/organizations/gh/{organization}'
+
+
+repo_re = re.compile(r'https://(www.)?github.com/([a-zA-Z0-9\-_]+)/([a-zA-Z0-9\-_]+)')
+RE_GROUP_ORG = 2
+RE_GROUP_REPO = 3
+
+SONAR_ORGS = {
```
Comment (zbedo) [inline security-controls/sast.py:46]: Can we avoid hardcoded values in the code?

File: security-controls/sast.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    gql = '''
+            query {
+                repository(name: $repo_name, owner: $org) {
+                    visibility
+                    isArchived
+                    primaryLanguage {
+                        name
+                    }
```
Comment (zbedo) [inline security-controls/sast.py:131]: Why is this necessary for SAST?

File: security-controls/sast.py (hunk 1 above)
Comment (TomConner) [reply security-controls/sast.py:12]: I want it for SAST management. It's well tested.

Comment (TomConner) [reply security-controls/sast.py:12]: Note I removed Codacy user info as discussed separately.

File: security-controls/deployment.yaml (hunk 6 above)
Comment (TomConner) [reply security-controls/deployment.yaml:40]: No need, that's harder to rotate should the need arise.

Comment (zbedo) [reply security-controls/deployment.yaml:40]: Why keeping keys in a separate file will help with rotation?

File: security-controls/sast.py (hunk 1 above)
Comment (zbedo) [reply security-controls/sast.py:12]: But I think if we are going to add SAST management will add it again, now we don't have it as part of this ticket/roadmap.

Comment (zbedo) [reply security-controls/sast.py:12]: That collection contains information for repositories that are not part of Terra services/security controls.

File: security-controls/sast.py (hunk 7 above)
Comment (TomConner) [reply security-controls/sast.py:52]: You make a good point about hardcoding. But the org list does not differ by environment and it seems better/simpler/easier to move it to the deploy script.

File: security-controls/sast.py (hunk 8 above)
Comment (TomConner) [reply security-controls/sast.py:46]: I'll move the ORGS lists out as discussed above. The regex and corresponding group numbers seems appropriately part of the code.

File: security-controls/sast.py (hunk 7 above)
Comment (zbedo) [reply security-controls/sast.py:52]: I understand, but since we want this project to be open-sourced, we should avoid hardcoding values.

File: security-controls/sast.py (hunk 9 above)
Comment (TomConner) [reply security-controls/sast.py:131]: SAST policy and tool support depends on visibility and code language.

File: docs/src/docs/secret_deplyment.md (hunk 2 above)
Comment (TomConner) [reply docs/src/docs/secret_deplyment.md:76]: Easier rotation and cleaner separation. Yes, I can follow the format.

File: sdarq/deploy.sh (hunk 3 above)
Comment (TomConner) [reply sdarq/deploy.sh:27]: It's there for reference in Firestore directly for now; we can expose this in the UI or reference it in policy enforcement code like the slack alerts as well.

File: security-controls/sast.py (hunk 4 above)
Comment (TomConner) [reply security-controls/sast.py:278]: To show if a secret was omitted or corrupt (without logging the secret itself, of course!). Now that it's working, I'll remove this debug message.

File: sdarq/deploy.sh (hunk 3 above)
Comment (zbedo) [reply sdarq/deploy.sh:27]: But we are not using it in the Sdarq namespace, if we are going to implement it in the future, we can add it.

File: security-controls/sast.py (hunk 5 above)
Comment (TomConner) [reply security-controls/sast.py:96]: We're only getting the metadata for repos with SAST running.

File: security-controls/deployment.yaml [hunk 10]
```diff
@@ -27,7 +28,7 @@ spec:
           runtimeClassName: gvisor
           nodeSelector:
             cloud.google.com/gke-nodepool: batch
-          restartPolicy: OnFailure
+          restartPolicy: Never
```
[the remark this answers is not in the document]
Comment (TomConner) [reply security-controls/deployment.yaml]: 👍

File: sdarq/deploy.sh (hunk 3 above)
Comment (TomConner) [reply sdarq/deploy.sh:27]: Removing from sdarq for now.

File: security-controls/deployment.yaml (hunk 6 above)
Comment (TomConner) [reply security-controls/deployment.yaml:40]: They're not really files, right? They're k8s secrets? It is necessary to know all the env vars within a secret in order to change any one of them because you appear to need to rewrite the whole secret. Maybe you have a simpler way -- I may have missed something in how to manage these.

File: security-controls/sast.py (hunk 7 above)
Comment (TomConner) [reply security-controls/sast.py:52]: You're right. I think firestore is probably the best place.

File: security-controls/sast.py (hunk 1 above)
Comment (TomConner) [reply security-controls/sast.py]: Moved to a separate branch.

**Comment by TomConner** [comment]
> For the `sast-details` collection that is added, as long as we don't need that information in this SAST update in security controls, I don't think we should include it in this PR. In the future, if we will need all that information we can add it again.

Moved to a different branch.

File: security-controls/sast.py (hunk 7 above)
Comment (TomConner) [reply security-controls/sast.py]: Added as build parameters.

**Review by zbedo — APPROVED** [review]

**Comment by zbedo** [comment]
This is great! Thank you!
